### ISSUE TITLE
upserting with 5MB range split size results in hung upserts

### ISSUE THREAD

**Issue by cody-yancey** [cockroachdb/cockroach#19797]
I'm doing some benchmarks by running a 16 node cluster in docker containers on Mesos (via Aurora).  

The image definition is:
```
FROM debian:jessie
ADD cockroach-v1.1.2.linux-amd64.tgz /usr/local/
RUN mv /usr/local/cockroach-*/* /usr/local/ && \
      rm -r /usr/local/cockroach-* && \
      apt-get update && \
      apt-get install -y netcat 
ADD run.sh /
```

The run script just waits until Instance 0 is running, and then joins it with:
```
    cockroach start \
        --insecure \
        --advertise-host=`hostname` \
        --cache=5GB \
        --http-port=$http \
        --join=$r0_host:$r0_rpc_port \
        --port=$rpc \
        --store=path=/mnt/mesos/sandbox/data,size=20GB
```

Instance 0 starts with:
```
    cockroach start \
        --insecure \
        --advertise-host=`hostname` \
        --cache=5GB \
        --http-port=$http  \
        --port=$rpc \
        --store=path=/mnt/mesos/sandbox/data,size=20GB
```
Where the ports are supplied by our software-networking layer.

I create a database using:
```
CREATE DATABASE jobs;
CREATE TABLE jobs.blob (id BIGINT PRIMARY KEY, data BYTES);
```
I then set the range split size to 5MB:

```
echo -e "range_min_bytes: 1\nrange_max_bytes: 5242880" | cockroach zone set .default --insecure -f - --host cockroachdb-test --port 81
```
Again, the host and port here are from our SWN stack.

I then load up the database with random blobs of data between 1K and 5K in size to simulate some of our wider tables. The keys are contiguous between 0 and 200,000 with no writes to the same key, in a random order, with a parallelization of 400, using a simple upsert statement:

```
UPSERT INTO jobs.blob(id, data) VALUES ($1, $2)
```
The admin UI reports 16 live nodes, a throughput of 30~ish writes per second, for a few minutes, before the cluster is completely hung (that is a separate bug I'll be filing). 

At this time, the admin UI reports that the blob table has 7 ranges, and 241.5 MiB of data, and `SHOW CLUSTER QUERIES` shows 399 hung UPSERT statements. Using some piped in SQL and xargs, I am able to cancel all of the UPSERT queries.

Then I go to SELECT from the table. There aren't that many entries at this point, so I try:
```
SELECT count(*) FROM jobs.blob
```
This query hangs indefinitely. [CTRL + C] cancels. Now `SHOW QUERIES` shows:
```
root@cockroachdb-test:81/> SHOW CLUSTER QUERIES;
+----------------------------------+---------+----------+----------------------------------+--------------------------------+-------------------+------------------+-------------+-----------+
|             query_id             | node_id | username |              start               |             query              |  client_address   | application_name | distributed |   phase   |
+----------------------------------+---------+----------+----------------------------------+--------------------------------+-------------------+------------------+-------------+-----------+
| 14f3c80c750b4966000000000000000c |      12 | root     | 2017-11-04 04:37:35.928042+00:00 | SHOW CLUSTER QUERIES           | ########:32966    | cockroach        | NULL        | preparing |
| 14f3c3abc77f2cc0000000000000000c |      12 | root     | 2017-11-04 03:17:22.653024+00:00 | SELECT count(*) FROM jobs.blob | ########:59998    | cockroach        | true        | executing |
+----------------------------------+---------+----------+----------------------------------+--------------------------------+-------------------+------------------+-------------+-----------+
(2 rows)

Time: 84.178846ms
```
Performing a CANCEL appears to succeed:
```
root@cockroachdb-test:81/> CANCEL QUERY '14f3c3abc77f2cc0000000000000000c';
CANCEL QUERY

Time: 77.902351ms
```
But subsequent calls to SHOW QUERY show that the query is still executing:

```
root@cockroachdb-test:81/> SHOW CLUSTER QUERIES;
+----------------------------------+---------+----------+----------------------------------+--------------------------------+-------------------+------------------+-------------+-----------+
|             query_id             | node_id | username |              start               |             query              |  client_address   | application_name | distributed |   phase   |
+----------------------------------+---------+----------+----------------------------------+--------------------------------+-------------------+------------------+-------------+-----------+
| 14f3c3abc77f2cc0000000000000000c |      12 | root     | 2017-11-04 03:17:22.653024+00:00 | SELECT count(*) FROM jobs.blob | ########:59998    | cockroach        | true        | executing |
| 14f3c84e26ae99a8000000000000000c |      12 | root     | 2017-11-04 04:42:18.081168+00:00 | SHOW CLUSTER QUERIES           | ########:32966    | cockroach        | NULL        | preparing |
+----------------------------------+---------+----------+----------------------------------+--------------------------------+-------------------+------------------+-------------+-----------+
(2 rows)

Time: 82.623762ms
```
CANCELing the query always reports success:
```
root@cockroachdb-test:81/> CANCEL QUERY '14f3c3abc77f2cc0000000000000000c';
CANCEL QUERY

Time: 124.662034ms

root@cockroachdb-test:81/> CANCEL QUERY '14f3c3abc77f2cc0000000000000000c';
CANCEL QUERY

Time: 126.849879ms

root@cockroachdb-test:81/> CANCEL QUERY '14f3c3abc77f2cc0000000000000000c';
CANCEL QUERY

Time: 86.815319ms

root@cockroachdb-test:81/> CANCEL QUERY '14f3c3abc77f2cc0000000000000000c';
CANCEL QUERY

Time: 133.631004ms

root@cockroachdb-test:81/> 
```

I would expect the query to no longer show up in the result of `SHOW QUERIES`.

All of the headers of the file look like this. As I am starting all of them in the exact same way, I'm fairly confident that giving **all 16 nodes** is unneeded.
```
Mesos Instance 0:
I171103 23:12:24.608236 1 util/log/clog.go:992  [config] file created at: 2017/11/03 23:12:24
I171103 23:12:24.608236 1 util/log/clog.go:992  [config] running on machine: ##########
I171103 23:12:24.608236 1 util/log/clog.go:992  [config] binary: CockroachDB CCL v1.1.2 (linux amd64, built 2017/11/02 19:32:03, go1.8.3)
I171103 23:12:24.608236 1 util/log/clog.go:992  [config] arguments: [cockroach start --insecure --advertise-host=######## --cache=5GB --http-port=31242 --port=31694 --store=path=/mnt/mesos/sandbox/data,size=20GB]

All Other Mesos Instances:
I171103 23:12:25.055112 1 util/log/clog.go:992  [config] file created at: 2017/11/03 23:12:25
I171103 23:12:25.055112 1 util/log/clog.go:992  [config] running on machine: ########
I171103 23:12:25.055112 1 util/log/clog.go:992  [config] binary: CockroachDB CCL v1.1.2 (linux amd64, built 2017/11/02 19:32:03, go1.8.3)
I171103 23:12:25.055112 1 util/log/clog.go:992  [config] arguments: [cockroach start --insecure --advertise-host=######## --cache=5GB --http-port=31459 --join=########:31694 --port=31922 --store=path=/mnt/mesos/sandbox/data,size=20GB]
```

**Comment by vivekmenezes** [comment]
@cody-yancey thank you for reporting this issue. I'm unable to reproduce it on a different setup and I'd appreciate your help reproducing it. I have a few questions.

- Is the cancel of the UPSERT queries related? I'd appreciate if you can run SELECT COUNT(*) before cancelling the pending UPSERT statements.
- Is setting the range split size required to reproduce this issue?

Thanks

**Comment by cody-yancey** [comment]
@vivekmenezes Thanks for the quick reply! I will try and find the rerpro case with the minimum set of steps sometime next week and get back to you. 

As for the state of the cluster right now, I can tell you that `SELECT COUNT(*)` always just hangs. I can `SELECT * FROM jobs.blob WHERE id > 0 AND id < 1000` but for certain ranges of id it will hang. I've been meaning to put together another bug report on that issue but it takes more time.

The cluster is currently hung right now (it is just an experiment I am running), so if there is any commands you want me run or logs you want me to send you I can certainly do that.

**Comment by vivekmenezes** [comment]
@cody-yancey I suspect your cluster has some unavailable ranges. Can you check if you can see any unavailable ranges on your cluster? You can see it on the admin UI at port 8080.

**Comment by cody-yancey** [comment]
The admin UI is actually on a random mesos port in my cluster, but it looks more or less the same on every node. No unavailable ranges.
![screen shot 2017-11-09 at 12 51 52 pm](https://user-images.githubusercontent.com/13629273/32626564-eb3f15e2-c54c-11e7-84fa-76bd91e13753.png)

**Comment by cody-yancey** [comment]
Another thing that I don't really understand, and this might be a potential clue, is that the replicas per node count keeps going up, long after writes have completely stopped.
![screen shot 2017-11-09 at 12 55 10 pm](https://user-images.githubusercontent.com/13629273/32626771-7be8f680-c54d-11e7-9051-412630b7ad77.png)

**Comment by asubiotto** [comment]
@cody-yancey I was able to reproduce a hung query after cancel and filed this issue to keep track of it: #19951

I'm not sure if the cause is the same though, so would you mind sharing a goroutine dump of the gateway node? (that would be node 12). You can access this by going to your admin ui and adding `/debug` to your address bar, then clicking on `All goroutines`.

In general, however, it seems that a `CANCEL QUERY` returning successfully does not imply that the query has been canceled (see https://github.com/cockroachdb/cockroach/blob/ef8bfb4b9dd364f9e60d5b8fa462ecb70c71634f/docs/RFCS/20170608_query_cancellation.md#new-cancel-query-statement). This should probably be clarified in the docs @jseldess.

**Comment by nstewart** [comment]
I'm having a similar issue. Here is the go-routine dump @asubiotto 
[long-running-query-goroutine.txt](https://github.com/cockroachdb/cockroach/files/1673882/long-running-query-goroutine.txt)

**Comment by jseldess** [comment]
@asubiotto, the [`CANCEL QUERY`](https://www.cockroachlabs.com/docs/stable/cancel-query.html#considerations) docs say this currently, under "Considerations":

> In rare cases where a query is close to completion when a cancellation request is issued, the query may run to completion.

That's the guidance I got when I wrote these docs. Should we expand this to cover how long-running queries may not actually get cancelled? Or is that a bug?

**Comment by asubiotto** [comment]
No, this is definitely a bug that wasn't fixed by #20534.

**Comment by asubiotto** [comment]
So I've seen pretty important delays in cancellation when doing some QA work for cancellation. I filed this issue as a result: https://github.com/cockroachdb/cockroach/issues/22650.

@cody-yancey, I think the reason that the cancellation is failing is due to your hung queries issue. that is a separate issue and should be investigated further. What happens with respect to cancellation is that because the query is hung, it does not produce a row, and unfortunately our current distsql infrastructure doesn't seem like it cancels a query until a first row is produced.

@nstewart, I believe your issue might also be related to the fact that cancellation didn't happen until a row is produced.

There are two separate issues here:
1) The hanging queries
2) Cancellation having no effect on queries that haven't produced rows.

I propose we rename this issue and focus on the hung queries. https://github.com/cockroachdb/cockroach/issues/22650 will serve to track the cancellation issue.

Re-assigning @vivekmenezes for triaging.

**Comment by vivekmenezes** [comment]
This issue is waiting on reproduction instructions. One issue that caused queries to hang has been fixed:  #19951

**Comment by cody-yancey** [comment]
After significant testing on Cockroach DB 2.0.0, I have been unable to reproduce this issue after almost a month of chaos and load testing. I strongly believe it has been resolved. Closing.
